File: code/ex993.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        if (root.val == x || root.val == y)    return false;
        
        Set<Integer> set = new HashSet<>();
        set.add(x);    set.add(y);
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean ifFind = false;
        
        while (!queue.isEmpty()) {
            for (int z = queue.size(); z > 0; --z) {
                TreeNode out = queue.poll();
                
                if (out.left != null && out.right != null) {    // Cannot have same parent
                    if (set.contains(out.left.val) && set.contains(out.right.val))
                        return false;
                }
                
                if (out.left != null) {
                    if (set.contains(out.left.val)) {
                        set.remove(out.left.val);
                        ifFind = true;
                    }
                    queue.offer(out.left);
                }
                
                if (out.right != null) {
                    if (set.contains(out.right.val)) {
                        set.remove(out.right.val);
                        ifFind = true;
                    }
                    queue.offer(out.right);
                }
            }
            
            if (ifFind)
                return set.size() == 0 ? true : false;
        }
        
        return false;
    }
}

File: code/ex993.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        int parent[101] = {0};
        
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty()) {
            int size = que.size();
            unordered_set<int> visited;
            for (int z = size; z > 0; z--) {
                TreeNode* out = que.front();    que.pop();
                if (out->left != NULL) {
                    visited.insert(out->left->val);
                    parent[out->left->val] = out->val;
                    que.push(out->left);
                }
                if (out->right != NULL) {
                    visited.insert(out->right->val);
                    parent[out->right->val] = out->val;
                    que.push(out->right);
                }
            }
            if (visited.find(x) != visited.end() && visited.find(y) != visited.end() && parent[x] != parent[y])
                return 1;
        }
        
        return 0;
    }
};